src/Contexts/PrintLoadingContext.tsx

```diff
@@ -1,40 +1,85 @@
+type AddPromiseDispatch = {type: PrintLoadingActions.ADD_PROMISE | PrintLoadingActions.ADD_PROMISE_INCR_EXPECTED, payload: Promise<unknown>}
+type ResetExpectedDispatch = {type: PrintLoadingActions.RESET_EXPECTED_COUNT}
+type PrintDispatch = {type: PrintLoadingActions.PRINT, expected: number}
+type PrintLoadingDispatch = AddExpectedDispatch | AddPromiseDispatch | ResetExpectedDispatch | PrintDispatch;
+
+const print = _.debounce(window.print, 3000, {leading: false, trailing: true});
+
+interface PrintLoadingState {
+    expected: number;
+    arr: Promise<unknown>[];
+}
 
 export enum PrintLoadingActions {
-    ADD_PROMISE
+    ADD_PROMISE,
+    ADD_PROMISE_INCR_EXPECTED,
+    ADD_EXPECTED_PROMISE_COUNT,
+    RESET_EXPECTED_COUNT,
+    PRINT,
 }
 
-const reducer = (state: Promise<unknown>[], action: PrintLoadingDispatch): Promise<unknown>[] => {
-    logger.debug('Adding a promise to the Loading context');
+const reducer = (state: PrintLoadingState, action: PrintLoadingDispatch): PrintLoadingState => {
     switch(action.type) {
+    case PrintLoadingActions.RESET_EXPECTED_COUNT:
+        logger.debug('Printing state reset.');
+        return {expected: 1, arr: []};
+    case PrintLoadingActions.ADD_EXPECTED_PROMISE_COUNT:
+        return {expected: state.expected + action.expected, arr: state.arr};
+    case PrintLoadingActions.ADD_PROMISE_INCR_EXPECTED:
+        logger.debug('Promise received');
+        // You must use the other action to increase expected counts.
+        return {expected: state.expected+1, arr: [...state.arr, action.payload]};
     case PrintLoadingActions.ADD_PROMISE:
-        return [...state, action.payload];
+        logger.debug('Promise received');
+        // You must use the other action to increase expected counts.
+        return {expected: state.expected, arr: [...state.arr, action.payload]};
+    case PrintLoadingActions.PRINT:
+        if (state.expected !== action.expected) {
+            logger.debug(`Skipping early print call from ${action.expected} -> ${state.expected}`);
+            return state;
+        }
+        logger.debug(`Printing because ${state.expected} == ${action.expected}`);
+        print();
+        return state;
     }
     return state;
 };
 
 export const usePrintLoadingContext = () => React.useContext(PrintLoadingContext);
 
 export const PrintLoadingProvider: React.FC<Props>  = ({ children }) => {
-    const [state, dispatch] = useReducer(reducer, []);
+    const [state, dispatch] = useReducer(reducer, {expected: 1, arr: []});
+
+    useEffect(()=>{
+        if (state.expected > 1 && state.arr.length >= state.expected && Promise.allSettled(state.arr)) {
+            logger.debug(`${state.expected} components have loaded, printing in two seconds.`);
+
+            setTimeout(() => {
+                dispatch({
+                    type: PrintLoadingActions.PRINT,
+                    expected: state.expected,
+                });
+            }, 10000);
```

Comment:
I thought you were working on getting rid of the 10 seconds wait? Is that coming later

Reply:
No, the 10 seconds is still going to stay to catch things we can't guarantee are tracked properly right now. These PRs address the 90% case where 10 seconds is too little time.